text/0003-font-types.md

```diff
@@ -0,0 +1,426 @@
+
+> **Q**: At what granularity should we think about copying costs? For instance,
+> I suspect that we do not want to copy bytes from tables that are unmodified.
+> Do we want to extend this to not copying bytes from individual subtables that
+> are unmodified, if other parts of a table are? Entries in subtables?
+
+The basic issue here is that if our main types are all just wrappers around raw
+bytes, we expect all of the components of a given object to be encoded
+contiguously; there is no way to say that "this subtable over here is in
+allocation A, and that one is reused from allocation B"; and each case where we
+want to handle that pattern will require some new type. At the very least we can
+imagine a special type for constructing a top-level font object (a header + a
+list of tables), where each table has its own `Buffer` object that may reuse
+storage from elsewhere; and we can then stream those bytes to disk or over the
+network without any additional storage. Doing this for each individual table,
+however, adds additional complexity.
+
+## Plain old data structures
+
+Another very compelling alternative is to just use plain old Rust data
+structures.
+
+In this world, our macros generate *two* types for each 'shape' that we define.
+One of these is the type that we've already seen, (methods for accessing data
+directly from underlying bytes) but the other one is just a plain old Rust
+struct.
+
+We can bikeshed names later; for the purpose of this illustration let's adopt
+the pattern of naming these new types objects `${TYPE}Owned` where `${TYPE}` is
+the name of the previously defined object. So continuing our `avar` example, we
+get:
+
+```rust
+// identical to above
+struct Avar<'a>(..);
+struct SegmentMaps<'a>(..);
+
+struct AvarOwned {
+    pub version_major: u16,
+    pub version_minor: u16,
+    pub segment_maps: Vec<SegmentMapsOwned>,
+}
+
+struct SegmentMapsOwned(pub Vec<AxisValueMap>);
+```
+
+All of these 'Owned' types would be plain Rust structs (and likely enums)
+generated by the macros, and they would also get generated code for writing them
+back out to bytes.
+
+This has the advantage of being a trivially easy API without any visible
+fanciness. You would describe your tables using simple rust types, using `std`
+APIs to mutate values, modify collections, or otherwise manipulate them. When
+you were happy with your representation, you would encode once, and then have
+the valid binary representation of your table (and you could still reuse the
+data structure, for instance modifying it further and writing out a different
+version).
+
+The downside of this is that there are now two types for each non-scalar
+structure in the spec; and additionally you are holding on to two
+representations of the data. In practice I don't mind this too much. One way to
+think about this is that these data structures are basically just another form
+of builder, that happen to be reusable and that maintain some intermediate
+state.
```

Comment:
We have this distinction already between fonttools-rs ("structures you want to manipulate") and otspec ("structures you want to serialize/deserialize") and it seems to work nicely. With this plan, the shaping engine can operate on "otspec"-level structures, and the font compiler/subsetter/etc. can continue to operate on "fonttools-rs"-level structures.

Reply:
yea, I think that's a good model; we would be combining those two approaches in a way where all the underlying scalar types were shared between both versions, and there would be easy conversion back and forth.